Source/Csla.AspNetCore/ApplicationContextManagerHttpContext.cs

```diff
@@ -113,44 +118,52 @@ public System.Security.Principal.IPrincipal GetUser()
     /// Sets the current principal.
     /// </summary>
     /// <param name="principal">Principal object.</param>
+    /// <exception cref="ArgumentNullException"><paramref name="principal"/> is <see langword="null"/>.</exception>
+    /// <exception cref="InvalidOperationException">The underlying <see cref="HttpContext"/> is <see langword="null"/>.</exception>
     public void SetUser(System.Security.Principal.IPrincipal principal)
     {
+      ArgumentNullException.ThrowIfNull(principal);
+      ThrowIfHttpContextIsNull();
       HttpContext.User = (ClaimsPrincipal)principal;
     }
 
     /// <summary>
     /// Gets the local context.
     /// </summary>
-    public IContextDictionary GetLocalContext()
+    public IContextDictionary? GetLocalContext()
     {
-      return (IContextDictionary)HttpContext?.Items[_localContextName];
+      return (IContextDictionary?)HttpContext?.Items[_localContextName];
     }
 
     /// <summary>
     /// Sets the local context.
     /// </summary>
     /// <param name="localContext">Local context.</param>
-    public void SetLocalContext(IContextDictionary localContext)
+    /// <exception cref="InvalidOperationException">The underlying <see cref="HttpContext"/> is <see langword="null"/>.</exception>
+    public void SetLocalContext(IContextDictionary? localContext)
     {
+      ThrowIfHttpContextIsNull();
       HttpContext.Items[_localContextName] = localContext;
     }
 
     /// <summary>
     /// Gets the client context.
     /// </summary>
     /// <param name="executionLocation"></param>
-    public IContextDictionary GetClientContext(ApplicationContext.ExecutionLocations executionLocation)
+    public IContextDictionary? GetClientContext(ApplicationContext.ExecutionLocations executionLocation)
     {
-      return (IContextDictionary)HttpContext?.Items[_clientContextName];
+      return (IContextDictionary?)HttpContext?.Items[_clientContextName];
     }
 
     /// <summary>
     /// Sets the client context.
     /// </summary>
     /// <param name="clientContext">Client context.</param>
     /// <param name="executionLocation"></param>
-    public void SetClientContext(IContextDictionary clientContext, ApplicationContext.ExecutionLocations executionLocation)
+    /// <exception cref="InvalidOperationException">The underlying <see cref="HttpContext"/> is <see langword="null"/>.</exception>
+    public void SetClientContext(IContextDictionary? clientContext, ApplicationContext.ExecutionLocations executionLocation)
     {
+      ThrowIfHttpContextIsNull();
```

Comment:
It should be valid to reset the value to null.

This class implements what amounts to a private API for use only by `ApplicationContext`, which provides the public API to the end developer.

From the context of a _private_ API, null should be a valid value, as it enjoys an optimization in serialization within the data portal.

Reply:
The `clientContext` itself can be `null` here. But I changed from a `NullReferenceException` to an `InvalidOperationException` because we _can't_ be sure `HttpContext` is not null here.